docs/site/pages/concepts/controllers.mdx

```diff
@@ -8,4 +8,245 @@ Head to the configuration reference section for [Controllers](/configuration/con
+        │                      │        [{                         
+        │     WASM Binary      │            "apiVersion": "..."    
+        │                      │            "kind": "...",         
+        │                      │            ...                    
+        │                      ├──────▶ }, ...]                    
+        └──────────────────────┘                                  
+```
+
+The purpose of this subcommand is to return a list of instances found by the target controller.
+The controller should handle filtering by namespace and optionall by a list of `key=value` pairs of labels.
+
+##### Output
+
+| Meaning   | Exit code | STDOUT                       |
+| --------- | --------- | ---------------------------- |
+| success   | 0         | JSON encoded resource stream |
+| error     | 1         | JSON encoded message         |
+| not found | 2         | JSON encoded message         |
+
+#### put
+
+Creating or updating an existing resource.
+
+```
+exec wasm ["put", "<kind>"]                                          
+                              ┌──────────────────────┐               
+{                             │                      │               
+    "apiVersion": "..."       │     WASM Binary      │               
+    "kind": "...",            │                      │               
+    ...                       │                      │               
+}                       ──────▶                      ├──────▶ { TBD }
+                              └──────────────────────┘               
+```
+
+The purpose of this subcommand is to create a new or update (upsert) an existing resource.
+Implementations should adjust the filesystem appropriately for the resource type and controllers needs.
+The new resource payload is serialized on STDIN.
+
+TBD:
+
+- What makes sense to return from the binary?
+
+##### Output
+
+| Meaning   | Exit code | STDOUT               |
+| --------- | --------- | -------------------- |
+| success   | 0         | TBD                  |
+| error     | 1         | JSON encoded message |
+
+#### delete
+
+Removing an existing resource.
+
+```
+exec wasm ["delete", "<kind>", "<namespace>", "<name>"]        
+        ┌──────────────────────┐                               
+        │                      │                               
+        │     WASM Binary      │                               
+        │                      │                               
+        │                      │                               
+        │                      ├──────▶ { TBD }                
+        └──────────────────────┘                               
+```
+
+The purpose of this subcommand is to remove an existing resource.
+Implementations should adjust the filesystem appropriately for the resource type and controllers needs.
+The namespace and name of the resource is passed as arguments to the subcommand.
+
+TBD:
+
+- What makes sense to return from the binary?
```

Comment:
We have a sample `wasm` controller written for Flipt resource definitions right? I wonder if it makes sense to link to that somewhere here for reference.

Reply:
That sounds like a great idea.

Reply:
I added a callout near the top of the WASM section. Let me know what you think 👍